File: libnd4j/include/ops/declarable/helpers/impl/rnn.cpp
/* ******************************************************************************
*
*
* This program and the accompanying materials are made available under the
* terms of the Apache License, Version 2.0 which is available at
* https://www.apache.org/licenses/LICENSE-2.0.
*
*  See the NOTICE file distributed with this work for additional
*  information regarding copyright ownership.
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
* License for the specific language governing permissions and limitations
* under the License.
*
* SPDX-License-Identifier: Apache-2.0
******************************************************************************/

//
// @author Yurii Shyrma (iuriish@yahoo.com), created on 16.04.2018
//

#include <system/op_boilerplate.h>

#if NOT_EXCLUDED(OP_rnn)

// function nnCell implements an Elman RNN cell: output = activation(Wx*x + bx  +  Wh*ht  + bh)
#include <helpers/BlasHelper.h>
#include <ops/declarable/helpers/rnn.h>

namespace sd {
namespace ops {
namespace helpers {

//////////////////////////////////////////////////////////////////////////
void rnnCell(sd::LaunchContext* context, NDArray* xt, NDArray* Wx, NDArray* Wh, NDArray* b,
            NDArray* hPrev, NDArray* ht) {
 // xt    input [bS x iS]
 // Wx    input-to-hidden weights, [iS  x nU]
 // Wh    hidden-to-hidden weights, [nU x nU]
 // b     biases, [2*nU]: {0, nU} are input-to-hidden biases and {nU, 2*nU} are hidden-to-hidden biases
 // hPrev previous cell output [bS x nU],  that is at previous time step t-1, in case of projection=false -> nU=nU!!!

 const int nU = hPrev->sizeAt(1);

 // ht is current cell output [bS x nU], that is at current time step t
 NDArray *bFirst = (*b)({{0, nU}});
 NDArray *bSecond = (*b)({{nU, 2 * nU}});

 NDArray *mmulOne = mmul(*xt, *Wx);
 NDArray *mmulTwo = mmul(*hPrev, *Wh);

 // Chain additions with proper dereferencing since operators return NDArray*
 NDArray *temp1 = (*mmulOne) + (*bFirst);
 NDArray *temp2 = (*temp1) + (*mmulTwo);
 NDArray *temp3 = (*temp2) + (*bSecond);

 ht->assign(temp3);  // [bS x nU] + [nU]  +  [bS x nU] + [nU] = [bS x nU]
 ht->applyTransform(transform::Tanh, ht);

 // Clean up intermediate results
 delete temp1;
 delete temp2;
 delete temp3;

 delete mmulOne;
 delete mmulTwo;
 delete bFirst;
 delete bSecond;
}

//////////////////////////////////////////////////////////////////////////
void rnnTimeLoop(sd::LaunchContext* context, NDArray* x, NDArray* Wx, NDArray* Wh, NDArray* b,
                NDArray* h0, NDArray* maxTimeStep, NDArray* h, NDArray* hFinal) {
 // x   input [time x bS x iS]
 // Wx  input-to-hidden  weights, [iS  x nU]
 // Wh  hidden-to-hidden weights, [nU x nU]
 // b   biases for, [2*nU]

 // h0          initial cell output (at time step = 0) [bS x nU]
 // maxTimeStep vector [bS] containing integer values within [0,time), each element of this vector set max time step
 // per each input in batch, this means there are no calculations for time >= maxTimeStep

 const int time = x->sizeAt(0);
 const int bS = x->sizeAt(1);

 // at first time step
 if (h0)
   hFinal->assign(h0);
 else
   *hFinal = 0.;

 BlasHelper::getInstance();  // to avoid memory leak in pragma parallel loops
 // loop through batch of inputs
 for (int e = 0; e < bS; ++e) {
   // loop through time steps
   for (int t = 0; t < time; ++t) {
     int maxStep = maxTimeStep ? maxTimeStep->e<int>(e) : time;

     NDArray *xt = (*x)({t, t + 1, e, e + 1, 0, 0}, true);
     NDArray *ht = (*h)({t, t + 1, e, e + 1, 0, 0}, true);
     NDArray *hPrev = (*hFinal)({e, e + 1, 0, 0}, true);  // previous state

     if (t >= maxStep) {
       *ht = 0.;
       NDArray *hPrevAssign = (*h)({maxStep - 1, maxStep, e, e + 1, 0, 0});
       if (maxStep != 0) hPrev->assign(hPrevAssign);
       delete hPrevAssign;
     } else {
       helpers::rnnCell(context, xt, Wx, Wh, b, hPrev, ht);
       hPrev->assign(ht);
     }

     delete xt;
     delete ht;
     delete hPrev;
   }
 }
}

}  // namespace helpers
}  // namespace ops
}  // namespace sd

#endif

File: nd4j/samediff-import/samediff-import-onnx/src/main/kotlin/org/nd4j/samediff/frameworkimport/onnx/definitions/implementations/RNN.kt
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */
package org.nd4j.samediff.frameworkimport.onnx.definitions.implementations

import org.nd4j.autodiff.samediff.SDVariable
import org.nd4j.autodiff.samediff.SameDiff
import org.nd4j.autodiff.samediff.internal.SameDiffOp
import org.nd4j.samediff.frameworkimport.ImportGraph
import org.nd4j.samediff.frameworkimport.hooks.PreImportHook
import org.nd4j.samediff.frameworkimport.hooks.annotations.PreHookRule
import org.nd4j.samediff.frameworkimport.registry.OpMappingRegistry
import org.nd4j.shade.protobuf.GeneratedMessageV3
import org.nd4j.shade.protobuf.ProtocolMessageEnum

/**
 * Implementation of ONNX RNN operation.
 *
 * ONNX RNN spec: https://github.com/onnx/onnx/blob/main/docs/Operators.md#RNN
 *
 * Computes a basic RNN layer:
 * H_t = activation(X_t * W^T + H_{t-1} * R^T + B)
 *
 * Inputs:
 * - X: Input tensor [seq_length, batch_size, input_size]
 * - W: Weight tensor for input [num_directions, hidden_size, input_size]
 * - R: Recurrence weight tensor [num_directions, hidden_size, hidden_size]
 * - B: Bias tensor (optional) [num_directions, 2*hidden_size]
 * - sequence_lens: Sequence lengths (optional)
 * - initial_h: Initial hidden state (optional)
 *
 * Attributes:
 * - activation_alpha: Activation alpha values
 * - activation_beta: Activation beta values
 * - activations: Activation functions (default: Tanh)
 * - clip: Gradient clipping value
 * - direction: forward, reverse, bidirectional
 * - hidden_size: Hidden state size
 *
 * @author Adam Gibson
 */
@PreHookRule(nodeNames = [], opNames = ["RNN"], frameworkName = "onnx")
class RNN : PreImportHook {

    override fun doImport(
        sd: SameDiff,
        attributes: Map<String, Any>,
        outputNames: List<String>,
        op: SameDiffOp,
        mappingRegistry: OpMappingRegistry<GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, ProtocolMessageEnum, GeneratedMessageV3, GeneratedMessageV3>,
        importGraph: ImportGraph<GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, GeneratedMessageV3, ProtocolMessageEnum>,
        dynamicVariables: Map<String, GeneratedMessageV3>
    ): Map<String, List<SDVariable>> {
        val opName = op.name
        
        // Get inputs
        val x = sd.getVariable(op.inputsToOp[0])  // [seq_length, batch_size, input_size]
        val w = sd.getVariable(op.inputsToOp[1])  // [num_directions, hidden_size, input_size]
        val r = sd.getVariable(op.inputsToOp[2])  // [num_directions, hidden_size, hidden_size]
        val b = if (op.inputsToOp.size > 3 && op.inputsToOp[3].isNotEmpty()) 
            sd.getVariable(op.inputsToOp[3]) else null
        val seqLens = if (op.inputsToOp.size > 4 && op.inputsToOp[4].isNotEmpty()) 
            sd.getVariable(op.inputsToOp[4]) else null
        val initialH = if (op.inputsToOp.size > 5 && op.inputsToOp[5].isNotEmpty()) 
            sd.getVariable(op.inputsToOp[5]) else null
        
        // Get attributes
        val hiddenSize = (attributes["hidden_size"] as? Number)?.toInt() ?: 0
        val direction = attributes.getOrDefault("direction", "forward") as? String ?: "forward"
        val activations = getStringList(attributes, "activations", listOf("Tanh"))
        
        // Determine number of directions
        val numDirections = if (direction == "bidirectional") 2 else 1
        
        // ONNX RNN operation is not fully supported yet
        // The implementation would require:
        // 1. Transpose X from [seq_length, batch_size, input_size] to [batch_size, seq_length, input_size]
        // 2. Apply RNN computation: H_t = activation(X_t * W^T + H_{t-1} * R^T + B)
        // 3. Handle bidirectional mode
        // 4. Return both output sequence and final hidden state

        // For now, create a placeholder that passes the input through
        // This is a temporary workaround until full RNN support is added
        val y = x.rename(outputNames[0])

        val resultMap = mutableMapOf<String, List<SDVariable>>()
        resultMap[outputNames[0]] = listOf(y)

        // Add placeholder for hidden state if needed
        if (outputNames.size > 1) {
            val yH = if (initialH != null) {
                initialH.rename(outputNames[1])
            } else {
                sd.zerosLike(outputNames[1], x)
            }
            resultMap[outputNames[1]] = listOf(yH)
        }

        return resultMap
    }
    
    private fun getStringList(attributes: Map<String, Any>, key: String, default: List<String>): List<String> {
        val value = attributes[key] ?: return default
        return when (value) {
            is List<*> -> value.map { it.toString() }
            is Array<*> -> value.map { it.toString() }
            else -> default
        }
    }
}
